File: src/blerpc_protocol/command.py
"""Command encode/decode layer for blerpc.

Command format (bits):
| type(1) | reserved(7) | cmd_name_len(8) | cmd_name(N*8) |
| data_len(16) | data(data_len*8) |

- type: 0=request, 1=response
- cmd_name: ASCII command name
- data_len: little-endian uint16
- data: protobuf-encoded bytes
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum


class CommandType(IntEnum):
    REQUEST = 0
    RESPONSE = 1


@dataclass
class CommandPacket:
    """A single command packet."""

    cmd_type: CommandType
    cmd_name: str
    data: bytes = b""

    def serialize(self) -> bytes:
        """Serialize command to bytes."""
        name_bytes = self.cmd_name.encode("ascii")
        if len(name_bytes) > 255:
            raise ValueError(f"cmd_name too long: {len(name_bytes)} > 255")
        if len(self.data) > 65535:
            raise ValueError(f"data too long: {len(self.data)} > 65535")

        # Byte 0: type in MSB (bit 7), reserved bits 6-0 = 0
        byte0 = (self.cmd_type & 0x01) << 7
        return (
            bytes([byte0])
            + struct.pack("<B", len(name_bytes))
            + name_bytes
            + struct.pack("<H", len(self.data))
            + self.data
        )

    @staticmethod
    def deserialize(data: bytes) -> CommandPacket:
        """Deserialize bytes into a CommandPacket."""
        if len(data) < 2:
            raise ValueError(f"Command packet too short: {len(data)} bytes")

        # Byte 0: type in MSB
        cmd_type = CommandType((data[0] >> 7) & 0x01)
        cmd_name_len = data[1]

        offset = 2
        if len(data) < offset + cmd_name_len + 2:
            raise ValueError("Command packet truncated")

        cmd_name = data[offset : offset + cmd_name_len].decode("ascii")
        offset += cmd_name_len

        data_len = struct.unpack_from("<H", data, offset)[0]
        offset += 2

        payload = data[offset : offset + data_len]
        return CommandPacket(
            cmd_type=cmd_type,
            cmd_name=cmd_name,
            data=payload,
        )

File: c/src/command.c
#include "blerpc_protocol/command.h"
#include <string.h>

int command_parse(const uint8_t *src, size_t src_len, struct command_packet *out)
{
    if (!src || !out || src_len < 2) {
        return -1;
    }

    /* Byte 0: type in MSB (bit 7) */
    out->cmd_type = (src[0] >> 7) & 0x01;
    out->cmd_name_len = src[1];

    size_t offset = 2;
    if (src_len < offset + out->cmd_name_len + 2) {
        return -1;
    }

    out->cmd_name = (const char *)(src + offset);
    offset += out->cmd_name_len;

    /* data_len: little-endian uint16 */
    out->data_len = (uint16_t)src[offset] | ((uint16_t)src[offset + 1] << 8);
    offset += 2;

    if (src_len < offset + out->data_len) {
        return -1;
    }

    out->data = src + offset;
    return 0;
}

int command_serialize(uint8_t cmd_type, const char *cmd_name, uint8_t cmd_name_len,
                      const uint8_t *data, uint16_t data_len, uint8_t *buf, size_t buf_size)
{
    size_t total = 2 + cmd_name_len + 2 + data_len;
    if (!buf || buf_size < total) {
        return -1;
    }

    /* Byte 0: type in MSB */
    buf[0] = (cmd_type & 0x01) << 7;
    buf[1] = cmd_name_len;

    if (cmd_name_len > 0 && cmd_name) {
        memcpy(buf + 2, cmd_name, cmd_name_len);
    }

    size_t offset = 2 + cmd_name_len;
    buf[offset] = (uint8_t)(data_len & 0xFF);
    buf[offset + 1] = (uint8_t)(data_len >> 8);
    offset += 2;

    if (data_len > 0 && data) {
        memcpy(buf + offset, data, data_len);
    }

    return (int)total;
}
